larq/quantized_variable.py

```diff
@@ -0,0 +1,308 @@
+"""Contains QuantizedVariable, a variable that can be quantized in the forward pass."""
+import tensorflow as tf
+from tensorflow.python.distribute.values import DistributedVariable
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import resource_variable_ops
+
+from larq import quantized_scope
+
+
+class QuantizedVariable(tf.Variable):
+    """A Variable that can be quantized in the forward pass in applicable contexts."""
+
+    def __init__(self, variable, quantizer=None, precision=None):
+        """Creates an QuantizedVariable instance.
+
+        # Arguments
+        variable: A floating-point resource variable to wrap.
+        quantizer: An optional quantizer to transform the floating-point
+            variable to a fake quantized variable.
+        precision: An optional integer defining the precision of the quantized
+            variable. If `None`, `quantizer.precision` is used.
+        """
+        if not resource_variable_ops.is_resource_variable(variable):
+            raise ValueError(
+                f"variable must be of type tf.ResourceVariable, but got: {variable}"
+            )
+        if not (quantizer is None or callable(quantizer)):
+            raise ValueError(
+                f"quantizer must be callable or None, but got: {quantizer}"
+            )
+        if not (precision is None or type(precision) == int):
+            raise ValueError(
+                f"precision must be of type integer or None, but got: {precision}"
+            )
+        self.latent_variable = variable
+        self.quantizer = quantizer
+        self.precision = precision or getattr(quantizer, "precision", None)
+
+    def _quantize(self, value):
+        if self.quantizer and quantized_scope.should_quantize():
+            return self.quantizer(value)
+        return value
+
+    def value(self):
+        return self._quantize(self.latent_variable.value())
+
+    def read_value(self):
+        return self._quantize(self.latent_variable.read_value())
+
+    def numpy(self):
+        return self._quantize(self.latent_variable).numpy()
+
+    def sparse_read(self, *args, **kwargs):
+        return self._quantize(self.latent_variable.sparse_read(*args, **kwargs))
+
+    def gather_nd(self, *args, **kwargs):
+        return self._quantize(self.latent_variable.gather_nd(*args, **kwargs))
+
+    def __getattr__(self, name):
+        return getattr(self.latent_variable, name)
```

Comment:
Following [this SO post](https://stackoverflow.com/a/1944708/1814949), should we do error handling for nonexistent names here?

Reply:
Not sure. I think it's fine not to handle any errors here, since `latent_variable` variable will throw an readable error message  if the attribute doesn't exist.